File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.cpp
class Solution {
public:
    int minCostConnectPoints(vector<vector<int>>& points) {
        int n = points.size();
        
        // min_heap to store minimum edge at top
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        pq.push({0, 0});
        int mstCost = 0, edgesUsed = 0;
        
        // Track nodes which are included in MST
        vector<bool> InMST(n);
        
        while (edgesUsed < n){
            auto [weight, currNode] = pq.top();
            pq.pop();
            
            // If node was already include in MST we will discard this edge
            if (InMST[currNode]) continue;
            InMST[currNode] = true;
            ++edgesUsed;
            mstCost += weight;
            
            for (int nextNode = 0; nextNode < n; ++nextNode){
                // If next node is not in MST, then edge from curr node
                // to next node can be pushed in the priority queue
                if (!InMST[nextNode]){
                    int nextWeight = abs(points[currNode][0] - points[nextNode][0]) + abs(points[currNode][1] - points[nextNode][1]);
                    pq.push({nextWeight, nextNode});
                }
            }
        }
        
        return mstCost;
    }
};

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        mst_cost, edges_used = 0, 0
        
        # Track nodes which are visited
        in_mst = [False] * n
        
        min_dist = [float('inf')] * n
        min_dist[0] = 0
        
        while edges_used < n:
            curr_min_edge = float('inf')
            curr_node = -1
            
            # pick least weight node which is not in MST
            for node in range(n):
                if not in_mst[node] and curr_min_edge > min_dist[node]:
                    curr_min_edge = min_dist[node]
                    curr_node = node
                
            mst_cost += curr_min_edge
            edges_used += 1
            in_mst[curr_node] = True
            
            # Update adjacent nodes of current node
            for next_node in range(n):
                weight = abs(points[curr_node][0] - points[next_node][0]) + abs(points[curr_node][1] - points[next_node][1])
                
                if not in_mst[next_node] and min_dist[next_node] > weight:
                    min_dist[next_node] = weight
            
        return mst_cost